boost-modular-build-helper] Add system library dl for boost-regex[icu] and boost-locale[icu] on linux

PhoebeHui opened the pull request:
Fixes https://github.com/microsoft/vcpkg/issues/16851

It requires system library 'dl' on linux when build with icu feature for boost-locale and boost-regex.

Failures:
```
gcc.link /home/josue/Projects/vcpkg/buildtrees/boost-locale/x64-linux-dbg/boost/build/807d196849ea9393cd67ccebca5daf06/has_icu64
/usr/bin/ld: /home/josue/Projects/vcpkg/installed/x64-linux/debug/lib/libicuuc.a(putil.ao): in function `uprv_dl_open_67':
/home/josue/Projects/vcpkg/buildtrees/icu/src/c-67_1-src-52188c6db3.clean/source/common/putil.cpp:2306: undefined reference to `dlopen'
/usr/bin/ld: /home/josue/Projects/vcpkg/installed/x64-linux/debug/lib/libicuuc.a(putil.ao): in function `uprv_dl_close_67':
/home/josue/Projects/vcpkg/buildtrees/icu/src/c-67_1-src-52188c6db3.clean/source/common/putil.cpp:2319: undefined reference to `dlclose'
/usr/bin/ld: /home/josue/Projects/vcpkg/installed/x64-linux/debug/lib/libicuuc.a(putil.ao): in function `uprv_dlsym_func_67':
/home/josue/Projects/vcpkg/buildtrees/icu/src/c-67_1-src-52188c6db3.clean/source/common/putil.cpp:2330: undefined reference to `dlsym'
collect2: error: ld returned 1 exit status
```

PhoebeHui committed:
[boost-modular-build-helper] Add system library dl for boost-regex[icu] and boost-locale[icu] on linux

PhoebeHui committed:
Update baseline version

PhoebeHui commented:
@josuegomes, could you help test this PR and see if it works for you?

cc @yurybura

josuegomes commented:
> @josuegomes, could you help test this PR and see if it works for you?
> 
> cc @yurybura

Still getting the same error.
I have applied the PR, cleaned up cache and buildtree, removed boost-locale, and tried again.
Log attached.
Is there anything I should look for on my environment?
[config.log](https://github.com/microsoft/vcpkg/files/6196701/config.log)

PhoebeHui commented:
@josuegomes, you need reinstall boost-modular-build-helper and build-build, you can remove it by execute './vcpkg remove uninstall-boost --recurse', it will remove all boost components.

I double confirmed locally, it looks fixed.

```
/home/phoebe/Hui/vcpkg$ cat ./buildtrees/boost-locale/x64-linux-dbg/boost/config.log
...found 15 targets...
...updating 5 targets...
gcc.compile.c++ /home/phoebe/Hui/vcpkg/buildtrees/boost-locale/x64-linux-dbg/boost/build/807d196849ea9393cd67ccebca5daf06/has_iconv_libc_obj.o
gcc.link /home/phoebe/Hui/vcpkg/buildtrees/boost-locale/x64-linux-dbg/boost/build/807d196849ea9393cd67ccebca5daf06/has_iconv
...updated 5 targets...
...found 63 targets...
...updating 2 targets...
gcc.compile.c++ /home/phoebe/Hui/vcpkg/buildtrees/boost-locale/x64-linux-dbg/boost/build/807d196849ea9393cd67ccebca5daf06/has_icu_obj.o
../src/../build/has_icu_test.cpp: In function ‘int main()’:
../src/../build/has_icu_test.cpp:22:12: warning: unused variable ‘c’ [-Wunused-variable]
    UChar32 c = ::u_charFromName(U_UNICODE_CHAR_NAME, "GREEK SMALL LETTER ALPHA", &err);
            ^
gcc.link /home/phoebe/Hui/vcpkg/buildtrees/boost-locale/x64-linux-dbg/boost/build/807d196849ea9393cd67ccebca5daf06/has_icu
...updated 2 targets...
...found 1 target...
...updating 1 target...
config-cache.write /home/phoebe/Hui/vcpkg/buildtrees/boost-locale/x64-linux-dbg/boost/project-cache.jam
```
